### ISSUE TITLE
odd shading around boxes

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#637]
See the boxes in

https://plus.google.com/u/0/+ShriramKrishnamurthi/posts/f33hP42DXR2

They look like they're somehow shaded. Actually, they're not. They were generated by

```
RECT-HEIGHT = 30
RECT-WIDTH = 30
FONT-SIZE = 10
BG = overlay(rectangle(RECT-WIDTH, RECT-HEIGHT, "outline", "black"),
  rectangle(RECT-HEIGHT, RECT-WIDTH, "solid", "white"))
SPACER = rectangle(RECT-WIDTH / 4, RECT-HEIGHT, "solid", "white")

fun as-image(ls :: List<String>) -> Image:
  cases (List) ls:
    | empty => SPACER
    | link(e, r) =>
      beside(SPACER, 
        beside(overlay(text(e, FONT-SIZE, "black"), BG),
          as-image(r)))
  end
end

fun words(w :: String) -> List<Image>:
  map(as-image, elt(w))
end
```

`BG` on its own is fine. `BG` with a letter is also fine. But the box when combined with the `SPACER` has this extra line.

![screen shot 2016-02-07 at 8 32 43 am](https://cloud.githubusercontent.com/assets/75030/12872639/82c880fa-cd75-11e5-833c-1af348bcb858.png)

**Comment by blerner** [comment]
This is confusing, but I think I've figured it out.  If I try a larger box size, on a plain white background, I get
![frame](https://cloud.githubusercontent.com/assets/918464/13899547/7a6fb9a6-edc7-11e5-8843-519272af428a.png)
It appears the `"outline"` style produces a anti-aliased frame, or maybe just a frame of width 2, or maybe just a frame of width 1 centered on a half-pixel-boundary, I'm not quite sure.  The reason the `BG` image alone looks fine is because that extra pixel is outside the boundary of the resulting image, so is cropped away, leaving the crisp line on three sides of your picture, but a thicker one on the right.  And the reason it's left-right asymmetric is because you're folding `beside` from the end of the list to the front, so you're overlapping one pixel of the frame with the `SPACER`, making for a thin left edge.

@schanzer do you see similar things with your image library in WeScheme (including with your latest improvements about anti-aliasing), or is this a bug we introduced somehow?

**Comment by schanzer** [comment]
@blerner, you're spot on with the explanation. The improvements I've made are only marginally better here, in that every *other* box has the extra pixel. I'll look into it.

**Comment by schanzer** [comment]
Yup - it's a sub-pixel issue. In fact, it's been plaguing WeScheme ever since it was first launched! @shriram , if you zoom in on the BG by itself, you'll find that it's _also_ rendering incorrectly! What's happening is that the outline is being drawn on the pixel boundary, so the browser anti-aliases a 2px-wide gray line. As @blerner points out, this gets cropped away in the first case. 

I _think_ the best way to address this is to make changes to each of the vertex-based primitives, so that they actually draw lines on pixels instead of pixel boundaries.  I've made a test case on WeScheme (http://www.wescheme.org/view?publicId=1y5LJn52q3) that reproduces the bug, and an experimental instance that addresses it for rectangles (http://subpixel.wescheme.appspot.com/openEditor?publicId=1y5LJn52q3).

**Comment by schanzer** [comment]
This is now fixed in WeScheme, though with one specific bit of hackery to deal with browser inconsistencies (details at https://github.com/bootstrapworld/wescheme2012/commit/d1f1d8f6f58bf01c5739fecb78bb4d844fa3eaac)

This fixes the specific bug @shriram reported, and also sharpens most primitives. Side-by-side comparison attached

![untitled](https://cloud.githubusercontent.com/assets/77126/14187847/05c8bbe4-f753-11e5-8d94-d06e211bc17f.png)

**Comment by schanzer** [comment]
Will be fixed once #670 is closed.
